### ISSUE TITLE
Protected fields: query constraints and write operations

### ISSUE THREAD

**Issue by BufferUnderflower** [parse-community/parse-server#6458]
**Is your feature request related to a problem? Please describe.**
1. Lack of functionality: as of now `protectedFields` do hide the field for `get`/`find` actions, but have no effect on `create`/`update`. 
2. Possible security issue: for `find` it is now possible to set query constraints using fields that are protected e.g.:
```js
{ 
  protectedFields: {
    '*': ['balance'] 
   }
}
...
object1.set('balance', 42);
object2.set('balance', 0);

const q = new Parse.Query(className);
q.greaterThanOrEqualTo('balance', 40);
const res = await q.find(); // results: [ object1 ]
```

As you can see there might be scenarios when this better work not.

**Describe the solution you'd like**
1. Skip the mutation of any protected fields while still mutating the ones w/o protection.
2. Execute the query ignoring the constraints accessing protected fields.

**Describe alternatives you've considered**
Alternative solution might be to throw an error instead of processing the request.

**Additional context**
I'm looking forward to implement it but not sure what behavior to stick with.

**Comment by BufferUnderflower** [comment]
### 1. Some points concerning `find`:

As of now if we add a constraint on a field that does not exist, the server still fulfills the query with constraint, as if the field's value was undefined, e.g. :

```js
// assuming we have an object:
// personA: { "balance": 100}

// query non-existing key:
query.greaterThan("deals", 0); // []

query.notEqual("deals", 0);   // [personA]
query.equalTo("deals", null);   // [personA]
query.equalTo("deals", undefined);   // [personA]
```

Handling is straightforward for rules defined at `*` / `authenticated` / `role:` / `id` scopes: given a user auth data(id and roles) we can calculate subset of fields he is not allowed to read and remove constraints on these fields from query before executing. 

This gets a bit tricky with rules defined by pointers: the fact that user has access to field is discovered during the query (depending on each object's content) - user might have access to protected fields on subset of the resulting objects (holding a pointer to this user), while for other objects he does not. The solution is to find out whether the pointer-protected fields are defined for this class before querying the database, and if so - extend the query with a condition. e.g.:
 
```js
{
 protectedFields:
 {
   "*": ["balance"],  // protected for everyone
   "userField:owner": [] // user pointed in 'owner' column can read all fields
 }
}
```

```js
// original query:
let query = { balance: {$gte: 0} }

// if there are pointer-protected field, becomes:
query = {
"$or": [
  // slice where field is allowed by pointer and constraint has effect
  {"$and": [ 
     { "balance": { "$gte": 0 } },
     { "_p_owner": "us3r1d" }
   ]
  },
  // remaining objects where field is not allowed for inspection
  // and constraint has no effect:
  { "_p_owner": { "$ne": "us3r1d" } }  
]}
```
So to achieve same behavior - split the query into multiple non-intersecting sets of results  (each object can't _have_ and _have not_ the pointer at the same time), and apply constraint for one part and do not for other. 

### 2. Concerning `create/update`: 
#### Current behavior:

When there's no field, it's added to schema (or request fails when `addField` is not permitted by CLP).

#### With protected fields:

1. Non-pointer: possible to remove protected fields from an `update` object (to mutate all other fields except protected).
2. Pointer-protected: this is what i'm currently stuck with. Parse is performing atomic `findOneAndUpdate`. Prior to mongo 4.2.0 there's no way to put a condition for `$set` operation. One solution is to split this query into separate find update: first check whether pointer is in object in find results, then decide what subset of fields should be protected and send appropriate update. I'm not sure if and how can this affect anything relying on atomic updates.


### To sum up: 
* Made it work for `find` and get expected results with any combinations of protectedFields and constraints.  No error thrown here. This may be reasonable in terms of flexibility (though i think "fail fast" approach could provide better developer experience, however this is a breaking change), so i choose to stick with _"pretend the field is undefined"_ behavior.
* Not sure whether to apply _"remove protected fields from update and process other fields"_ or _"cancel such update throwing an error"_

@davimacedo do you have any thoughts?
